### PR TITLE
DOC: Improved docs for jnp.fft.rfft and irfft

### PR DESCRIPTION
**Pull request by rajasekharporeddy**
Part of #21461

### CODE REVIEW

File: jax/_src/numpy/fft.py [hunk 1]
```diff
@@ -400,15 +400,133 @@ def ifft(a: ArrayLike, n: int | None = None,
...
+     [12.+0.j   -3.+1.73j]]
+
+    When ``n=4`` and ``axis=0``, dimension of the transform along ``axis 0`` will
+    be ``(4/2)+1 =3`` and dimension along other axes will be same as that of input.
+
+    >>> with jnp.printoptions(precision=2, suppress=True):
+    ...   print(jnp.fft.rfft(x, n=4, axis=0))
+    [[ 3.+0.j  7.+0.j 11.+0.j 15.+0.j]
+     [ 1.-2.j  3.-4.j  5.-6.j  7.-8.j]
+     [-1.+0.j -1.+0.j -1.+0.j -1.+0.j]]
```
Comment (dfm) [inline jax/_src/numpy/fft.py:460]: It looks like this output isn't quite correct. It looks like the final column should be removed.

File: jax/_src/numpy/fft.py [hunk 2]
```diff
@@ -400,15 +400,133 @@ def ifft(a: ArrayLike, n: int | None = None,
...
+  r"""Compute a one-dimensional discrete Fourier transform for real input.
+
+  JAX implementation of :func:`numpy.fft.rfft`.
+
+  Args:
+    a: real-valued input array.
+    n: int. Specifies the dimension of the result along ``axis``. The dimension
+      of the array along ``axis`` is ``(n/2)+1``, if ``n`` is even and ``(n+1)/2``,
+      if ``n`` is odd. If not specified, ``n`` defaults to the dimension of ``a``
+      along ``axis``.
```
Comment (dfm) [inline jax/_src/numpy/fft.py:415]: It might be easier to think about this as the number of inputs that are used, rather than related to the dimension of the output. Right now, I think it's a bit confusing to say that `n` "specifies" the output dimension, when the output dimension is a function of `n`.

File: jax/_src/numpy/fft.py [hunk 3]
```diff
@@ -400,15 +400,133 @@ def ifft(a: ArrayLike, n: int | None = None,
...
+    Array(True, dtype=bool)
+  """
   return _fft_core_1d('rfft', xla_client.FftType.RFFT, a, n=n, axis=axis,
                       norm=norm)
 
-@implements(np.fft.irfft)
+
 def irfft(a: ArrayLike, n: int | None = None,
           axis: int = -1, norm: str | None = None) -> Array:
+  r"""Compute a one-dimensional inverse discrete Fourier transform for real input.
```
Comment (dfm) [inline jax/_src/numpy/fft.py:475]: Reword similar to `rfft`.

File: jax/_src/numpy/fft.py [hunk 4]
```diff
@@ -400,15 +400,133 @@ def ifft(a: ArrayLike, n: int | None = None,
...
+
+  JAX implementation of :func:`numpy.fft.irfft`.
+
+  Args:
+    a: real-valued input array.
+    n: int. Specifies the dimension of the result along ``axis``. If not specified,
+      the dimension of the result along ``axis`` is ``2*(m-1)``, where ``m`` is the
```
Comment (dfm) [inline jax/_src/numpy/fft.py:482]:
```suggestion
      the dimension of the result along ``axis`` is ``n = 2*(m-1)``, where ``m`` is the
```

It might be worth adding this ⬆️ or something similar because I got confused about the relationship between `n` and `m`.

**Review by dfm — CHANGES_REQUESTED** [review]
Thanks! I left some small comments inline. Can you also rebase onto the current `main` to fix unrelated linter errors?

File: jax/_src/numpy/fft.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/numpy/fft.py:415]: Is it more accurate to say that `n` specifies the effective input size?

File: jax/_src/numpy/fft.py (hunk 3 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/fft.py:474]: If reword this similar to the suggestion for `rfft`, then it will exceed 82 characters which goes to second line. But Jake suggests that we should try to make the one-line summary fit in one line. Can I make the change as suggested or can I keep it as it is? Please suggest.

File: jax/_src/numpy/fft.py (hunk 2 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/fft.py]: Thanks! Changed the description of `n` as `n: int. Specifies the effective dimension of the input along ``axis``. If not specified, it will default to the dimension of input along ``axis``.` and the discussion of output size is moved to "Returns:".

File: jax/_src/numpy/fft.py (hunk 1 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/fft.py]: Done

File: jax/_src/numpy/fft.py [hunk 5]
```diff
@@ -400,15 +400,133 @@ def ifft(a: ArrayLike, n: int | None = None,
   return _fft_core_1d('ifft', xla_client.FftType.IFFT, a, n=n, axis=axis,
                       norm=norm)
 
-@implements(np.fft.rfft)
+
 def rfft(a: ArrayLike, n: int | None = None,
          axis: int = -1, norm: str | None = None) -> Array:
+  r"""Compute a one-dimensional discrete Fourier transform for real input.
```
[the remark this answers is not in the document]
Comment (rajasekharporeddy) [reply jax/_src/numpy/fft.py]: Done

File: jax/_src/numpy/fft.py (hunk 4 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/fft.py]: Changed

File: jax/_src/numpy/fft.py (hunk 3 above)
Comment (dfm) [reply jax/_src/numpy/fft.py:474]: Good question! I'm happy to keep this as is to stay within the line length limit. Thanks!

**Comment by dfm** [comment]
Thanks @rajasekharporeddy! It looks like there is an issue with the example code: it's failing with a dimension mismatch error. Can you take a look? Otherwise looks great!

**Comment by rajasekharporeddy** [comment]
Thanks @dfm, I have fixed the indentation in code example. The outputs in code examples are checked and they look fine. I don't know why its failing with a dimension mismatch error. Can you please run the tests once again and check if it still failing with the same error.

**Comment by rajasekharporeddy** [comment]
Sorry, I got what the error here. Here we can not reproduce `x` using `jnp.fft.irfft(jnp.fft.rfft(x)`. I will remove that code.

**Comment by rajasekharporeddy** [comment]
Rebased onto latest `main` and squashed all commits into one.

**Review by dfm — APPROVED** [review]
Looks good - thanks!
